src/coap_io.c

```diff
@@ -850,9 +850,10 @@ coap_network_read(coap_socket_t *sock, coap_packet_t *packet) {
 #else
       if (errno == ECONNREFUSED) {
 #endif
-	/* client-side ICMP destination unreachable, ignore it */
-	coap_log(LOG_WARNING, "coap_network_read: unreachable\n");
-	return -2;
+        /* client-side ICMP destination unreachable, ignore it */
```

Comment:
A UDP socket should be closed when the client session ends, because despite the error it is still valid. The -2 return value already guarantees the session state is reset.

Reply:
If the target server is not available, but trying to set up a DTLS session, and a -2 is returned from coap_network_read, the logic in coap_read_session() calls coap_session_reset() but never calls coap_socket_close() leaving the socket file descriptor still open.

If an application keeps on trying to call coap_new_client_session_psk() - not just the once as coap-client does, and you look at the file descriptors, you will see them increasing - as :

# ls -l /proc/19691/fd
total 0
lrwx------ 1 root root 64 Feb 12 19:38 0 -> /dev/pts/0
lrwx------ 1 root root 64 Feb 12 19:38 1 -> /dev/pts/0
lrwx------ 1 root root 64 Feb 12 19:39 10 -> socket:[11118524]    <== 
lrwx------ 1 root root 64 Feb 12 19:39 11 -> socket:[11118774]
lrwx------ 1 root root 64 Feb 12 19:40 12 -> socket:[11119012]
lrwx------ 1 root root 64 Feb 12 19:41 13 -> socket:[11119208]
lrwx------ 1 root root 64 Feb 12 19:38 2 -> /dev/pts/0
lrwx------ 1 root root 64 Feb 12 19:38 3 -> socket:[11118042]
.. Additional other fds.

# netstat -anp | grep client
udp        0      0 192.168.0.189:37381         192.168.191.2:5684          ESTABLISHED 19691/./client    
udp        0      0 192.168.0.189:50195         192.168.191.2:5684          ESTABLISHED 19691/./client
udp        0      0 192.168.0.189:35091         192.168.191.2:5684          ESTABLISHED 19691/./client  
udp        0      0 192.168.0.189:40860         192.168.191.2:5684          ESTABLISHED 19691/./client  
udp        0      0 192.168.0.189:56481         192.168.191.2:5684          ESTABLISHED 19691/./client  
udp        0      0 192.168.0.189:42532         192.168.191.2:5684          ESTABLISHED 19691/./client  
udp        0      0 192.168.0.189:33191         192.168.191.2:5684          ESTABLISHED 19691/./client  
udp        0      0 192.168.0.189:54443         192.168.191.2:5684          ESTABLISHED 19691/./client  
udp        0      0 192.168.0.189:44846         192.168.191.2:5684          ESTABLISHED 19691/./client  
udp        0      0 192.168.0.189:45871         192.168.191.2:5684          ESTABLISHED 19691/./client  
udp        0      0 192.168.0.189:48725         192.168.191.2:5684          ESTABLISHED 19691/./client  
udp        0      0 192.168.0.189:50531         192.168.191.2:5684          ESTABLISHED 19691/./client

Reply:
By design the number of open sockets will increase if you create multiple client sessions to the same server, but this isn’t a leak. Presumably your would release the previous session before you create a new one if your application logic calls for only one socket. Unlike server sessions, client sessions always have to be actively managed by the client, regardless of the state of the session, since only the client knows if it needs to keep the session opened or if it is done with it.
In this particular case (datagram with icmp not reachable) there is no reason to close the socket. You can send a new message right away on the same socket and it may or may not succeed, the error is not sticky.